src/hotspot/share/opto/loopnode.cpp

```diff
@@ -5150,21 +5152,40 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
         set_loop(C->root(), _ltree_root);
       }
     }
-    // Weeny check for irreducible.  This child was already visited (this
-    // IS the post-work phase).  Is this child's loop header post-visited
-    // as well?  If so, then I found another entry into the loop.
-    if (!_verify_only) {
+    if (is_postvisited(l->_head)) {
+      // We are currently visiting l, but its head has already been post-visited.
+      // l is irreducible: we just found a second entry m.
+      _has_irreducible_loops = true;
+      RegionNode* secondary_entry = m->as_Region();
+      assert(secondary_entry != nullptr, "irreducible loop secondary_entry is Region");
+      assert(secondary_entry->loop_status() == RegionNode::LoopStatus::MaybeIrreducibleEntry,
+             "secondary_entry has wrong loop_status");
+      assert(!secondary_entry->is_Loop(), "LoopNode cannot be secondary entry to irreducible loop");
+
+      // Walk up the loop-tree, mark all loops that are already post-visited as irreducible
+      // Since m is a secondary entry to them all.
       while( is_postvisited(l->_head) ) {
-        // found irreducible
         l->_irreducible = 1; // = true
+        RegionNode* head = l->_head->as_Region();
+        assert(head != nullptr, "irreducible loop head is Region");
+        assert(head->loop_status() == RegionNode::LoopStatus::MaybeIrreducibleEntry,
+               "head has wrong loop_status");
+        assert(!head->is_Loop(), "LoopNode cannot be irreducible loop head");
```

Comment:
It is the same asserts as above. May be move to one function.

Reply:
Ok, I will move it to a function `RegionNode::as_IrreducibleLoopEntry()`.